doc/api/esm.md

```diff
@@ -945,31 +915,61 @@ export function resolve(specifier, context, defaultResolve) {
   return defaultResolve(specifier, context, defaultResolve);
 }
 
-export function getFormat(url, context, defaultGetFormat) {
+export async function load(url, context, defaultLoad) {
   // Now that we patched resolve to let CoffeeScript URLs through, we need to
-  // tell Node.js what format such URLs should be interpreted as. For the
-  // purposes of this loader, all CoffeeScript URLs are ES modules.
+  // tell Node.js what format such URLs should be interpreted as. Because
+  // CoffeeScript transpiles into JavaScript, it should be one of the two
+  // JavaScript formats: 'commonjs' or 'module'.
   if (extensionsRegex.test(url)) {
+    // CoffeeScript files can be either CommonJS or ES modules, so we want any
+    // CoffeeScript file to be treated by Node.js the same as a .js file at the
+    // same location. To determine how Node.js would interpret an arbitrary .js
+    // file, search up the file system for the nearest parent package.json file
+    // and read its "type" field.
+    const format = await getPackageType(url);
+    // When a hook returns a format of 'commonjs', `source` is be ignored.
+    // To handle CommonJS files, a handler needs to be registered with
+    // `require.extensions` in order to process the files with the CommonJS
+    // loader. Avoiding the need for a separate CommonJS handler is a future
+    // enhancement planned for ES module loaders.
+    if (format === 'commonjs') {
+      return { format };
+    }
+
+    const { source: rawSource } = await defaultLoad(url, { format });
+    // This hook converts CoffeeScript source code into JavaScript source code
+    // for all imported CoffeeScript files.
+    const transformedSource = CoffeeScript.compile(rawSource.toString(), {
+      bare: true,
+      filename: url,
+    });
+
     return {
-      format: 'module'
+      format,
+      source: transformedSource,
     };
   }
 
   // Let Node.js handle all other URLs.
-  return defaultGetFormat(url, context, defaultGetFormat);
+  return defaultLoad(url, context, defaultLoad);
 }
 
-export function transformSource(source, context, defaultTransformSource) {
-  const { url, format } = context;
+async function getPackageType(url) {
+  const isFile = !!extname(url);
```

Comment:
Oof. darn, good catch. I think it'll take an actual FS hit then, which is way more expensive 😥

Reply:
looking it also would fail for query params since extname is just for file paths, not urls

Reply:
Do we need to check whether it’s a file or folder? What if we just tried to read it in `try/catch` and if it succeeds, it’s a file (and presumably it’s the `package.json` we’re looking for) else it’s presumed to be a folder and we keep looking?

Reply:
Actually, I think it doesn't matter: package.json is not extension-less. What if I changed the constant's name and add a comment noting the caveat?

Reply:
Are you saying that the only file you care about is `package.json`? So then rather than checking if `url` is a file, why not just check specifically if it’s `package.json`?

```suggestion
  const isPackageJson = url.endsWith('package.json');
```

And replace the lower reference to `isFile` with `isPackageJson`.

Reply:
Be wary here, `path.extname('file:///root/home.js?page=/ ')` for example returns `''`

Reply:
@bmeck it fails because of the trailing slash (otherwise, it returns `'.js?page='`), which won't be present in the resolved url. If a user choses to add it before passing to getPackageType, well, they shot themselves in the foot.

Reply:
I'm not sure I understand, this would affect any valid url with a trailing `/` in the fragment it doesn't seem to be a user choice as the loader doesn't control what things an application may import.

Reply:
Sorry, my previous message had a typo that I corrected: I wrote "trailing slash" but meant "trailing space". The slash without the trailing space is okay:

```js
path.extname('file:///root/home.js?page=/') // '.js?page=/'
```

Reply:
I mean, there are all sorts of exceptions we could concoct (e.g. `path.extname('file:///root/node.js/')`) around this but the most robust fix is to just try to load the file like @GeoffreyBooth mentioned regardless of your or my string shenanigans.

Reply:
Yees, it COULD do an fs.read on `url` and check the `dirent.isFile()`, but I think for an example, such fool-proofing is overkill. If this were brought into node-core, then it would seem appropriate to catch the edge cases (perhaps the fs.read is not the silver-bullet, and it needs the query params stripped or something else).

Reply:
i'd agree it doesn't need to be perfect, so how it is seems sufficient without becoming so verbose as to be confusing

Reply:
This is just an example in the docs, not part of core, so we don’t have to obsess over performance. I think an extra file system call is okay if that makes the example easier to read.

If there’s something that’s both readable _and_ robust that would be ideal, I think. Maybe something like “try to read `` `${url}/package.json` ``, if that throws try to read `url`, else recurse with `dirname(url)`”?

Reply:
Sure, that would work too.

I went with the extname check because it's comparatively very cheap and pretty robust. For instance, with the query param, the trailing space character is internally passed through URL(), which would convert it to an encoded url entity, which extname handles fine too.

My thinking was that people will probably copy+paste the example, and as-is, it will work in almost all cases. So then almost everyone gets the more efficient version. Someone with an edge-case didn't get there by accident, so they probably have a good idea of what's happening and what to do (plus the code comments I added warning the the file may not exist etc).

Off the top of my head, I can't think of a gotcha to the fs.read approach as you described @GeoffreyBooth, but there probably is at least one too.

I can switch it over, but tonight is my last night before I'm on holiday for a week, so if we wanna land this PR before then, that's all the time left.

Reply:
We could do the safer version (two FS calls) and add a comment like “If you’re sure your URL doesn’t contain spaces or encoded characters, you could replace this first call with `extname(url)` and get a performance improvement” or whatever. That way we cover all (?) cases by default, and people are tipped off to the potential performance improvement if they know that the edge cases don’t apply to them.

Reply:
Sure. I'll try to find time tonight to update and test.

Reply:
@GeoffreyBooth as it turns out, this is way more complicated than simply adding an extra fs read. Manipulating the paths becomes a snake-pit of "what kind of path is this", "that kind of path doesn't work with this utility", and having to catch/handle more and more expected errors. I haven't even covered all the failure scenarios for the node-loaders, and `getPackageType()` has already tripled in complexity.

I really think the existing example is sufficient, and the edge-cases are probably better handled upstream anyway.

Reply:
Okay, can we perhaps just add a comment then? Something like “This simple check fails on edge cases such as extensionless files or URLs with spaces; but is sufficient for most projects” or whatever it should say. Then we’ve done our part in terms of writing a good example, or at least one without a known/undeclared bug, and someone else can take on the challenge of improving the example in a future PR.